### ISSUE TITLE
oxfmt: /*@__PURE__*/ position rewritten outside parens, losing rolldown tree-shaking hint

### ISSUE THREAD

**Issue by jdalton** [oxc-project/oxc#22795]
## Summary

oxfmt rewrites `/*@__PURE__*/` (and related rolldown/rollup/terser pure-annotation comments) into a position where rolldown silently ignores them. The reformatted output is valid JavaScript but loses its tree-shaking semantics.

## Reproduce

Input (oxfmt 1.x, current main):

```ts
function getMatchesGlob() {
  if (!probed) {
    const fn = (
      /*@__PURE__*/ require('node:path') as typeof import('node:path') & {
        matchesGlob?: unknown
      }
    ).matchesGlob
    return fn
  }
}
```

Run with the minimal config:

```json
{
  "tabWidth": 2,
  "printWidth": 80,
  "singleQuote": true,
  "trailingComma": "all",
  "semi": false
}
```

Output (oxfmt rewrites the comment outside the parens):

```ts
function getMatchesGlob() {
  if (!probed) {
    const fn = /*@__PURE__*/ (
      require('node:path') as typeof import('node:path') & {
        matchesGlob?: unknown
      }
    ).matchesGlob
    return fn
  }
}
```

## Expected

The `/*@__PURE__*/` comment should stay adjacent to the `require()` call expression (i.e. immediately to its left, inside the parens). When it gets hoisted outside the parenthesized cast, rolldown's pure-annotation detector treats the parenthesized expression as the annotation target (not the inner call) and silently drops the annotation. The call ends up not marked pure, and dead-code elimination misses it.

## Why this matters

The rolldown docs are explicit about position requirements: https://rolldown.rs/in-depth/dead-code-elimination#pure. The same is true for rollup, terser, and esbuild — all use position-adjacent matching for the magic-comment family.

Affected annotations (all positional):

- `/*@__PURE__*/` and `/* @__PURE__ */`
- `/*#__PURE__*/` and `/* #__PURE__ */`
- `/*@__NO_SIDE_EFFECTS__*/` and `/* @__NO_SIDE_EFFECTS__ */`
- `/*#__NO_SIDE_EFFECTS__*/` and `/* #__NO_SIDE_EFFECTS__ */`

## Workaround

Restructure to an intermediate const so the cast sits on its own statement:

```ts
const pathMod = /*@__PURE__*/ require('node:path') as typeof import('node:path') & {
  matchesGlob?: unknown
}
const fn = pathMod.matchesGlob
```

oxfmt handles this idempotently — line-reflows the cast but doesn't move the comment.

## Suggested fix

In oxfmt's comment-attachment phase, treat the magic-comment family above as anchored to the immediately-following AST node (a CallExpression in the `/*@__PURE__*/` case, a FunctionDeclaration / ClassDeclaration in the `/*@__NO_SIDE_EFFECTS__*/` case). Skip the usual leading-comment-hoisting rule for those.

## Repro environment

- oxfmt: latest (reproduced locally on 2026-05-28)
- Node: 26.1.0
- Platform: darwin-arm64

Happy to send a PR with a regression test + the comment-anchor patch if the maintainers agree on the approach.

**Comment by leaysgur** [comment]
NOTE to self: Prettier also formats in the same way.

https://playground.oxc.rs/?t=formatter&formatterPanels=output%2Cprettier&options=%7B%22run%22%3A%7B%22lint%22%3Atrue%2C%22formatter%22%3Atrue%2C%22transform%22%3Afalse%2C%22isolatedDeclarations%22%3Afalse%2C%22whitespace%22%3Afalse%2C%22mangle%22%3Afalse%2C%22compress%22%3Afalse%2C%22scope%22%3Atrue%2C%22symbol%22%3Atrue%2C%22cfg%22%3Afalse%7D%2C%22parser%22%3A%7B%22extension%22%3A%22tsx%22%2C%22allowReturnOutsideFunction%22%3Atrue%2C%22preserveParens%22%3Atrue%2C%22allowV8Intrinsics%22%3Atrue%2C%22semanticErrors%22%3Atrue%7D%2C%22linter%22%3A%7B%7D%2C%22formatter%22%3A%7B%22useTabs%22%3Afalse%2C%22tabWidth%22%3A2%2C%22endOfLine%22%3A%22lf%22%2C%22printWidth%22%3A80%2C%22singleQuote%22%3Afalse%2C%22jsxSingleQuote%22%3Afalse%2C%22quoteProps%22%3A%22as-needed%22%2C%22trailingComma%22%3A%22all%22%2C%22semi%22%3Atrue%2C%22arrowParens%22%3A%22always%22%2C%22bracketSpacing%22%3Atrue%2C%22bracketSameLine%22%3Afalse%2C%22objectWrap%22%3A%22preserve%22%2C%22singleAttributePerLine%22%3Afalse%7D%2C%22transformer%22%3A%7B%22target%22%3A%22es2015%22%2C%22useDefineForClassFields%22%3Atrue%2C%22experimentalDecorators%22%3Atrue%2C%22emitDecoratorMetadata%22%3Atrue%2C%22optimizeEnums%22%3Atrue%2C%22optimizeConstEnums%22%3Atrue%7D%2C%22isolatedDeclarations%22%3A%7B%22stripInternal%22%3Afalse%7D%2C%22codegen%22%3A%7B%22normal%22%3Atrue%2C%22jsdoc%22%3Atrue%2C%22annotation%22%3Atrue%2C%22legal%22%3Atrue%7D%2C%22compress%22%3A%7B%7D%2C%22mangle%22%3A%7B%22topLevel%22%3Atrue%2C%22keepNames%22%3Afalse%7D%2C%22controlFlow%22%3A%7B%22verbose%22%3Afalse%7D%2C%22inject%22%3A%7B%22inject%22%3A%7B%7D%7D%2C%22define%22%3A%7B%22define%22%3A%7B%7D%7D%7D&code=function+getMatchesGlob%28%29+%7B%0A++if+%28%21probed%29+%7B%0A++++const+fn+%3D+%28%0A++++++%2F*%40__PURE__*%2F+require%28%27node%3Apath%27%29+as+typeof+import%28%27node%3Apath%27%29+%26+%7B%0A++++++++matchesGlob%3F%3A+unknown%0A++++++%7D%0A++++%29.matchesGlob%0A++++return+fn%0A++%7D%0A%7D

**Comment by sapphi-red** [comment]
```js
const fn = /*@__PURE__*/ (
  foo()
).foo;
const fn2 = (
  /*@__PURE__*/ foo()
).foo;
```
Both `foo()` are detected as `pure` by Oxc ([playground](https://playground.oxc.rs/?t=ir&options=%7B%22run%22%3A%7B%22lint%22%3Afalse%2C%22formatter%22%3Afalse%2C%22transform%22%3Afalse%2C%22isolatedDeclarations%22%3Afalse%2C%22whitespace%22%3Afalse%2C%22mangle%22%3Afalse%2C%22compress%22%3Afalse%2C%22scope%22%3Atrue%2C%22symbol%22%3Atrue%2C%22cfg%22%3Afalse%7D%2C%22parser%22%3A%7B%22extension%22%3A%22ts%22%2C%22allowReturnOutsideFunction%22%3Atrue%2C%22preserveParens%22%3Atrue%2C%22allowV8Intrinsics%22%3Atrue%2C%22semanticErrors%22%3Atrue%7D%2C%22linter%22%3A%7B%7D%2C%22formatter%22%3A%7B%22useTabs%22%3Afalse%2C%22tabWidth%22%3A2%2C%22endOfLine%22%3A%22lf%22%2C%22printWidth%22%3A80%2C%22singleQuote%22%3Afalse%2C%22jsxSingleQuote%22%3Afalse%2C%22quoteProps%22%3A%22as-needed%22%2C%22trailingComma%22%3A%22all%22%2C%22semi%22%3Atrue%2C%22arrowParens%22%3A%22always%22%2C%22bracketSpacing%22%3Atrue%2C%22bracketSameLine%22%3Afalse%2C%22objectWrap%22%3A%22preserve%22%2C%22singleAttributePerLine%22%3Afalse%7D%2C%22transformer%22%3A%7B%22target%22%3A%22es2015%22%2C%22useDefineForClassFields%22%3Atrue%2C%22experimentalDecorators%22%3Atrue%2C%22emitDecoratorMetadata%22%3Atrue%2C%22optimizeEnums%22%3Atrue%2C%22optimizeConstEnums%22%3Atrue%7D%2C%22isolatedDeclarations%22%3A%7B%22stripInternal%22%3Afalse%7D%2C%22codegen%22%3A%7B%22normal%22%3Atrue%2C%22jsdoc%22%3Atrue%2C%22annotation%22%3Atrue%2C%22legal%22%3Atrue%7D%2C%22compress%22%3A%7B%7D%2C%22mangle%22%3A%7B%22topLevel%22%3Atrue%2C%22keepNames%22%3Afalse%7D%2C%22controlFlow%22%3A%7B%22verbose%22%3Afalse%7D%2C%22inject%22%3A%7B%22inject%22%3A%7B%7D%7D%2C%22define%22%3A%7B%22define%22%3A%7B%7D%7D%7D&code=const+fn+%3D+%2F*%40__PURE__*%2F+%28%0A++foo%28%29%0A%29.foo%3B%0Aconst+fn2+%3D+%28%0A++%2F*%40__PURE__*%2F+foo%28%29%0A%29.foo%3B%0A)).


> The call ends up not marked pure, and dead-code elimination misses it.

Do you have a repro for this? I wonder if it should be fixed on the Rolldown side.
